### PR TITLE
[Colocate plan][Step1] Colocate join covers more situations

### LINKED ISSUE #5184: cocolate join在查询中带函数的情况下会失效
同一个表包含聚合函数的查询时，也必须要join，但是却不是cocolate join，总是报"colocate: false, reason: Node type not match"，为什么会这样，能优化吗？
比如下列语句：
select count() from (select user_id from t_event where  event_name = 'AppOpen' group by user_id having count() > 5) t1 where user_id in (select user_id from t_event where  event_name = 'Buy' group by user_id having count() > 3)

### LINKED ISSUE #5589: Colocate plan
**Is your feature request related to a problem? Please describe.**

Currently Doris supports colocate table building and colocate join. 
But it did not make full use of the advantages of colocate.
For example, when there is an Aggregation Node between Hash Join Node and Scan Node, colocate join cannot be performed.
For another example, Aggregation Node, Sort Node, and Set Operation nodes can also be absorbed by child nodes when the data distribution matches, but they are not currently available.

**Describe the solution you'd like**

In the case of reasonable data distribution, the upper-level operators can be absorbed by the lower-level operators, 
thereby reducing unnecessary network transmission and serialization and deserialization operations.

**Describe alternatives you've considered**

The main operators for colocate optimization include the following 4 plan node:
Hash Join Node
Set Operation Node
Aggregation Node
Sort Node (in window function)

Step1: create table
```
CREATE TABLE `test_colocate` (
  `k1` int(11) NULL COMMENT "",
  `k2` int(11) NULL COMMENT "",
  `k3` int(11) NULL COMMENT "",
  `k4` int(11) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`k1`, `k2`, `k3`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k1`, `k2`) BUCKETS 10
PROPERTIES (
"replication_num" = "2",
"in_memory" = "false",
"storage_format" = "V2"
);
```

Step2:  colocate join

```
explain select * from (select k1, k2 from test_colocate group by k1, k2) a , test_colocate b where a.k1=b.k1 and a.k2=b.k2;
+-------------------------------------------------------------------------------------------------------------------------+
| Explain String                                                                                                          |
+-------------------------------------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                                                         |
|  OUTPUT EXPRS:<slot 2> `k1` | <slot 3> `k2` | `b`.`k1` | `b`.`k2` | `b`.`k3` | `b`.`k4`                                 |
|   PARTITION: UNPARTITIONED                                                                                              |
|                                                                                                                         |
|   RESULT SINK                                                                                                           |
|                                                                                                                         |
|   4:EXCHANGE                                                                                                            |
|                                                                                                                         |
| PLAN FRAGMENT 1                                                                                                         |
|  OUTPUT EXPRS:                                                                                                          |
|   PARTITION: HASH_PARTITIONED: `default_cluster:test`.`test_colocate`.`k1`, `default_cluster:test`.`test_colocate`.`k2` |
|                                                                                                                         |
|   STREAM DATA SINK                                                                                                      |
|     EXCHANGE ID: 04                                                                                                     |
|     UNPARTITIONED                                                                                                       |
|                                                                                                                         |
|   3:HASH JOIN                                                                                                           |
|   |  join op: INNER JOIN                                                                                                |
|   |  hash predicates:                                                                                                   |
|   |  colocate: true                                                                                                     |
|   |  equal join conjunct: <slot 2> `k1` = `b`.`k1`                                                                      |
|   |  equal join conjunct: <slot 3> `k2` = `b`.`k2`                                                                      |
|   |                                                                                                                     |
|   |----2:OlapScanNode                                                                                                   |
|   |       TABLE: test_colocate                                                                                          |
|   |                                                                                                                     |
|   1:AGGREGATE (update finalize)                                                                                         |
|   |  group by: `k1`, `k2`                                                                                               |
|   |                                                                                                                     |
|   0:OlapScanNode                                                                                                        |
|      TABLE: test_colocate                                                                                               |
+-------------------------------------------------------------------------------------------------------------------------+
47 rows in set (0.020 sec)

```

Step3: colocate aggregation node
Condition: The input partition of aggregation node >= the data partition of child fragment
```
 explain select k1, k2  from test_colocate where k1=1  group by k1, k2;
+-------------------------------------------------------------------------------------------------------------------------+
| Explain String                                                                                                          |
+-------------------------------------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                                                         |
|  OUTPUT EXPRS:<slot 2> `k1` | <slot 3> `k2`                                                                             |
|   PARTITION: UNPARTITIONED                                                                                              |
|                                                                                                                         |
|   RESULT SINK                                                                                                           |
|                                                                                                                         |
|   2:EXCHANGE                                                                                                            |
|                                                                                                                         |
| PLAN FRAGMENT 1                                                                                                         |
|  OUTPUT EXPRS:                                                                                                          |
|   PARTITION: HASH_PARTITIONED: `default_cluster:test`.`test_colocate`.`k1`, `default_cluster:test`.`test_colocate`.`k2` |
|                                                                                                                         |
|   STREAM DATA SINK                                                                                                      |
|     EXCHANGE ID: 02                                                                                                     |
|     UNPARTITIONED                                                                                                       |
|                                                                                                                         |
|   1:AGGREGATE (update finalize)                                                                                         |
|   |  group by: `k1`, `k2`                                                                                               |
|   |                                                                                                                     |
|   0:OlapScanNode                                                                                                        |
|      TABLE: test_colocate                                                                                               |
+-------------------------------------------------------------------------------------------------------------------------+
30 rows in set (0.011 sec)
```

Step4: sort node (in window function)
Condition:
The sort columns of sort node >= the data partition of child fragment
```
The sort column is k1 and k2.

explain select k1, sum(k2) over(partition by k1 order by k2) from test_colocate;
+-------------------------------------------------------------------------------------------------------------------------+
| Explain String                                                                                                          |
+-------------------------------------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                                                         |
|  OUTPUT EXPRS:<slot 4> <slot 0> | <slot 3>                                                                              |
|   PARTITION: UNPARTITIONED                                                                                              |
|                                                                                                                         |
|   RESULT SINK                                                                                                           |
|                                                                                                                         |
|   3:EXCHANGE                                                                                                            |
|                                                                                                                         |
| PLAN FRAGMENT 1                                                                                                         |
|  OUTPUT EXPRS:                                                                                                          |
|   PARTITION: HASH_PARTITIONED: `default_cluster:test`.`test_colocate`.`k1`, `default_cluster:test`.`test_colocate`.`k2` |
|                                                                                                                         |
|   STREAM DATA SINK                                                                                                      |
|     EXCHANGE ID: 03                                                                                                     |
|     UNPARTITIONED                                                                                                       |
|                                                                                                                         |
|   2:ANALYTIC                                                                                                            |
|   |  functions: [, sum(<slot 5> ), ]                                                                                    |
|   |  partition by: `k1`                                                                                                 |
|   |  order by: <slot 5>  ASC                                                                                            |
|   |  window: RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW                                                          |
|   |                                                                                                                     |
|   1:SORT                                                                                                                |
|   |  order by: <slot 4> <slot 0> ASC, <slot 5>  ASC                                                                     |
|   |  offset: 0                                                                                                          |
|   |                                                                                                                     |
|   0:OlapScanNode                                                                                                        |
|      TABLE: test_colocate                                                                                               |
+-------------------------------------------------------------------------------------------------------------------------+
36 rows in set (0.028 sec)
```

Colocate set operation node
Condition1: The data partition of all child fragments >= input partition of set operation node
Condition2: If there is no exchange node in child fragment, it will mean that the data hasn't been rehashed.
Condition3: Scan node are all in the same colocate group

step1: create table
```
CREATE TABLE `t1` (
    `id` int(11) COMMENT "",
    `value` varchar(8) COMMENT ""
    ) ENGINE=OLAP
    DUPLICATE KEY(`id`)
    DISTRIBUTED BY HASH(`id`) BUCKETS 10
    PROPERTIES (
    "colocate_with" = "t1",
    "replication_num" = "2"
    );

    CREATE TABLE `t2` (
    `id` int(11) COMMENT "",
    `value` varchar(8) COMMENT ""
    ) ENGINE=OLAP
    DUPLICATE KEY(`id`)
    DISTRIBUTED BY HASH(`id`) BUCKETS 10
    PROPERTIES (
    "colocate_with" = "t1",
    "replication_num" = "2"
    );

    CREATE TABLE `t3` (
    `id` int(11) COMMENT "",
    `value` varchar(8) COMMENT ""
    ) ENGINE=OLAP
    DUPLICATE KEY(`id`)
    DISTRIBUTED BY HASH(`id`) BUCKETS 10
    PROPERTIES (
    "colocate_with" = "t1",
    "replication_num" = "2"
    );
```

step2: colocate query

The t1, t2, t3 are in the same colocate group and the intersect column `id` is same as the distributed column `id`.

```
explain select id from t1 intersect select id from t2 intersect select id from t3;
+----------------------------------------------------------------------------------+
| Explain String                                                                   |
+----------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                  |
|  OUTPUT EXPRS:<slot 3> `id` `id` `id`                                            |
|   PARTITION: UNPARTITIONED                                                       |
|                                                                                  |
|   RESULT SINK                                                                    |
|                                                                                  |
|   4:EXCHANGE                                                                     |
|                                                                                  |
| PLAN FRAGMENT 1                                                                  |
|  OUTPUT EXPRS:                                                                   |
|   PARTITION: HASH_PARTITIONED: `default_cluster:test`.`t1`.`id`                  |
|                                                                                  |
|   STREAM DATA SINK                                                               |
|     EXCHANGE ID: 04                                                              |
|     UNPARTITIONED                                                                |
|                                                                                  |
|   0:INTERSECT                                                                    |
|   |  colocate=true                                                               |
|   |                                                                              |
|   |----2:OlapScanNode                                                            |
|   |       TABLE: t2                                                              |
|   |                                                                              |
|   |----3:OlapScanNode                                                            |
|   |       TABLE: t3                                                              |
|   |                                                                              |
|   1:OlapScanNode                                                                 |
|      TABLE: t1                                                                   |
+----------------------------------------------------------------------------------+
```

**Additional context**

The colocate plan will also have some bad cases. 
For example, when the data is skewed, because colocate eliminates the problem of rehash, the query will be more affected by the data skew.
In this case, it is best to give priority to solving the problem of data skew and set a reasonable fragmentation column.
Of course, you can also turn off colocate through session configuration.

### PR DESCRIPTION
**Pull request by EmmyMiao87**
## Proposed changes
There are 4 steps in colocate plan: join, agg, sort and set operation.
This is the first step.

The old colocate join can only cover the case where the child is hash or scan.
In fact, as long as the child's data distribution meets the requirements,
  no matter what the plan node on the child node is, a colocate join can be performed.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #5589) and described the bug/feature there in detail. Later
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

document is comming

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/system/SystemInfoService.java [hunk 1]
```diff
@@ -44,6 +41,10 @@
 import com.google.common.collect.Multimap;
 import com.google.common.collect.Sets;
 
+import org.apache.commons.validator.routines.InetAddressValidator;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
```
Comment (HappenLee) [inline fe/fe-core/src/main/java/org/apache/doris/system/SystemInfoService.java:46]: some as up comment

File: fe/fe-core/src/test/java/org/apache/doris/planner/ColocatePlanTest.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    @Test
+    public void sqlDistributedSmallerThanData2() throws Exception {
+        String sql = "explain select * from (select k1 from db1.test_colocate group by k1, k2) a , db1.test_colocate b "
+                + "where a.k1=b.k1";
+        String plan1 = UtFrameUtils.getSQLPlanOrErrorMsg(ctx, sql);
+        Assert.assertTrue(plan1.contains(DistributedPlanColocateRule.INCONSISTENT_DISTRIBUTION_OF_TABLE_AND_QUERY));
+    }
+
+    // with:
+    // 1. agg columns = distributed columns
```
Comment (HappenLee) [inline fe/fe-core/src/test/java/org/apache/doris/planner/ColocatePlanTest.java:79]: group by columns is better?

File: fe/fe-core/src/test/java/org/apache/doris/planner/ColocatePlanTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        String sql = "explain select * from (select k1, k2 from db1.test_colocate group by k1, k2) a , db1.test_colocate b "
+                + "where a.k1=b.k1 and a.k2=b.k2";
+        String plan1 = UtFrameUtils.getSQLPlanOrErrorMsg(ctx, sql);
+        Assert.assertEquals(1, StringUtils.countMatches(plan1, "AGGREGATE"));
+        Assert.assertTrue(plan1.contains(COLOCATE_ENABLE));
+    }
+
+    // with:
+    // 1. agg columns > distributed columns
+    // 2. hash columns = agg output columns > distributed columns
```
Comment (HappenLee) [inline fe/fe-core/src/test/java/org/apache/doris/planner/ColocatePlanTest.java:92]: hash columns? means join columns?

File: fe/fe-core/src/main/java/org/apache/doris/planner/PlanNode.java [hunk 4]
```diff
@@ -30,15 +30,15 @@
 import org.apache.doris.thrift.TPlan;
 import org.apache.doris.thrift.TPlanNode;
 
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
-
```
[the remark this answers is not in the document]
Comment (EmmyMiao87) [reply fe/fe-core/src/main/java/org/apache/doris/planner/PlanNode.java:35]: The Doris class should be at the top of Imports.

The Logger class belongs to others import. So it should be under the com.google.common

File: fe/fe-core/src/main/java/org/apache/doris/system/SystemInfoService.java (hunk 1 above)
Comment (EmmyMiao87) [reply fe/fe-core/src/main/java/org/apache/doris/system/SystemInfoService.java:46]: same as above comment

File: fe/fe-core/src/test/java/org/apache/doris/planner/ColocatePlanTest.java (hunk 3 above)
Comment (EmmyMiao87) [reply fe/fe-core/src/test/java/org/apache/doris/planner/ColocatePlanTest.java:92]: Yes

**Comment by EmmyMiao87** [comment]
> Why change import order of LogManager?

The order of import should comply with http://doris.apache.org/master/zh-CN/developer-guide/fe-eclipse-dev.html#%E4%BB%A3%E7%A0%81%E6%9B%B4%E6%96%B0

**Comment by EmmyMiao87** [comment]
> seems not need return null too many, may under code is better:
> 
> ```
>        if (this instanceof ScanNode && tupleIds.contains(tupleId)) {
>             return (ScanNode) this;
>         } else if (!(this instanceof ExchangeNode)) {
>             for (PlanNode planNode : children) {
>                 ScanNode scanNode = planNode.getScanNodeInOneFragmentByTupleId(tupleId);
>                 if (scanNode != null) {
>                     return scanNode;
>                 }
>             }
>         }
>         return null;
> ```

changed

**Comment by EmmyMiao87** [comment]
> the comment is same as line 570. just keep one is enough.

removed

File: fe/fe-core/src/main/java/org/apache/doris/planner/PlanNode.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (HappenLee) [reply fe/fe-core/src/main/java/org/apache/doris/planner/PlanNode.java]: OK

**Comment by HappenLee** [comment]
LGTM

**Review by HappenLee — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/system/SystemInfoService.java [hunk 5]
```diff
@@ -781,11 +782,16 @@ public void releaseBackends(String clusterName, boolean isReplay) {
...
         }
 
+
         // if more than one backend exists in same host, select a backend at random
         List<Backend> backends = Lists.newArrayList();
         for (List<Backend> list : backendMaps.values()) {
-            Collections.shuffle(list);
-            backends.add(list.get(0));
+            if (FeConstants.runningUnitTest) {
+                backends.addAll(list);
```
[the remark this answers is not in the document]
Comment (EmmyMiao87) [reply fe/fe-core/src/main/java/org/apache/doris/system/SystemInfoService.java:790]: If it is a test, add all the backends belonging to this ip.

File: fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 6]
```diff
@@ -294,16 +298,52 @@ private boolean isBroadcastCostSmaller(long broadcastCost, long partitionCost)
...
+     * and transform it into PlanFragment.
      */
     private PlanFragment createHashJoinFragment(HashJoinNode node, PlanFragment rightChildFragment,
                                                 PlanFragment leftChildFragment, long perNodeMemLimit,
                                                 ArrayList<PlanFragment> fragments)
             throws UserException {
+        List<String> reason = Lists.newArrayList();
+        if (canColocateJoin(node, leftChildFragment, rightChildFragment, reason)) {
+            node.setColocate(true, "");
+            //node.setDistributionMode(HashJoinNode.DistributionMode.PARTITIONED);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java:315]: remove unused code

File: fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 7]
```diff
@@ -471,58 +468,167 @@ private PlanFragment createHashJoinFragment(HashJoinNode node, PlanFragment righ
...
-            // Before we support global runtime filter, only shuffle join do not enable local runtime filter
+            // TODO: Before we support global runtime filter, only shuffle join do not enable local runtime filter
             node.setIsPushDown(false);
             return joinFragment;
         }
     }
 
+    /**
+     * Colocate Join can be performed when the following 4 conditions are met at the same time.
+     * 1. Session variables disable_colocate_plan = true
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java:479]:
```suggestion
     * 1. Session variables disable_colocate_plan = false
```

File: fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 8]
```diff
@@ -471,58 +468,167 @@ private PlanFragment createHashJoinFragment(HashJoinNode node, PlanFragment righ
...
+            return null;
+        }
+        ScanNode scanNode = planFragment.getPlanRoot()
+                .getScanNodeInOneFragmentByTupleId(slotRef.getDesc().getParent().getId());
+        if (scanNode == null) {
+            cannotReason.add(DistributedPlanColocateRule.REDISTRIBUTED_SRC_DATA);
+            return null;
         }
+        if (scanNode instanceof OlapScanNode) {
+            return ((OlapScanNode) scanNode);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java:539]:
```suggestion
            return (OlapScanNode) scanNode;
```

File: fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 9]
```diff
@@ -812,6 +843,7 @@ private PlanFragment createSetOperationNodeFragment(
         }
 
         // There is at least one partitioned child fragment.
+        // TODO(ML): here
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java:846]: TODO what？

File: fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 10]
```diff
@@ -932,9 +964,9 @@ private PlanFragment createAggregationFragment(
         if (isDistinct) {
             return createPhase2DistinctAggregationFragment(node, childFragment, fragments);
         } else {
-
             // Check table's distribution. See #4481.
             PlanNode childPlan = childFragment.getPlanRoot();
+            // TODO(ML): here
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java:969]: ？

File: fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 11]
```diff
@@ -1175,8 +1207,9 @@ private PlanFragment createAnalyticFragment(
             // required if the sort partition exprs reference a tuple that is made nullable in
             // 'childFragment' to bring NULLs from outer-join non-matches together.
             DataPartition sortPartition = sortNode.getInputPartition();
+            // TODO(ML): here
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java:1210]: ？

File: fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java (hunk 9 above)
Comment (EmmyMiao87) [reply fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java:846]: Used to mark the location of subsequent colocate plan work. But it has now been developed. no need

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**